crates/core/src/db/update.rs

```diff
@@ -125,19 +145,67 @@ pub fn schema_updates(
     for proposed_schema_def in proposed_tables {
         let proposed_table_name = &proposed_schema_def.table_name;
         if let Some(known_schema) = known_tables.remove(proposed_table_name) {
-            // Not pretty, but roundtripping through `TableDef` ensures all the
-            // fields are ordered consistently.
-            // Also resets ids to zero.
-            let known_schema = TableSchema::from_def(known_schema.table_id, TableDef::from(known_schema.into_owned()));
+            // Unfortunately `TableSchema::from_def . TableDef::from != id`.
+            //
+            // Also, there is no guarantee that the constituents of the schema
+            // are sorted. They will be, however, when converting the proposed
+            // `TableDef` into `TableSchema` (via `from_def`).
+            let known_schema = known_schema.into_owned();
+            let columns = known_schema
+                .columns()
+                .iter()
+                .cloned()
+                .sorted_by_key(|x| x.col_pos)
+                .collect();
```

Comment:
It looks to me like the only change necessary is to sort the columns here, all the other stuff is done in `TableSchema::from_def`, so  I propose that instead, we create `TableSchema:sort_columns(mut self)` that also recompute the `row_type` and keep the rest the same:

```rust
fn compute_row_type(columns: &[ColumnSchema]) -> ProductType {
        ProductType::new(
            columns
                .iter()
                .map(|c| ProductTypeElement {
                    name: Some(c.col_name.clone()),
                    algebraic_type: c.col_type.clone(),
                })
                .collect(),
        )
    }

    pub fn sort_columns(&mut self) {
        self.columns.sort_by_key(|x| x.col_pos);
        self.row_type = Self::compute_row_type(&self.columns);
    }
```

then is only:

```rust
        if let Some(known_schema) = known_tables.remove(proposed_table_name) {
            let mut known_schema =
                TableSchema::from_def(known_schema.table_id, TableDef::from(known_schema.into_owned()));
            known_schema.sort_columns();

            let proposed_schema = TableSchema::from_def(known_schema.table_id, proposed_schema_def);

            if proposed_schema != known_schema {
```

Reply:
That is not correct, from_def also adds additional “generated” definitions and doesn’t correctly remove them.

Please feel free to fix the datatypes instead.

Reply:
Do you mean that in `from_def` the `generated` get duplicates from the ones supplied by `TableDef`? 


Anyway, if is the case then approve this PR and later fix it

Reply:
Yes, `from_def` assumes that the given `TableDef` cannot know about those generated values (because the module cannot know about them). However, converting an existing `TableSchema` into a `TableDef` will leave the generated ones in place. So if we turn that into a `TableSchema` again, there will be duplicates.

I added this to the docstring.

It is hard to tell which conversions should be legal -- my expectation would be that `TableSchema -> TableDef` should "forget" about anything a `TableDef` coming from a module would not contain. I would also expect that `TableSchema::new` ensures the same ordering as `TableSchema::from_def`.

It is also odd that all collections are `Vec`s, when semantically they should be sets. I suppose that's because what is stored in the system tables is canonical ultimately.

One approach could be to make `validated` correct all of those potential pitfalls, and return a newtype. So we know we're comparing `Validated(TableSchema) == Validated(Tableschema)`.